File: 25.reverse-nodes-in-k-group/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	ListNode* reverseKGroup(ListNode* head, int k) {
		if (!head)
		{
			return nullptr;
		}
		ListNode* lastNode = nullptr;
		ListNode* p = head;
		while (true)
		{
			if (!p) return head;
			ListNode* subHead = p;
			for (int i = 0; i < k - 1; i++)
			{
				p = p->next;
				if (!p) return head;
			}
			ListNode* tempNext = p->next;
			p->next = nullptr;
			ListNode* newEnd;
			if (!lastNode) head = swapAll(subHead, newEnd);
			else lastNode->next = swapAll(subHead, newEnd);
			lastNode = newEnd;
			p = newEnd->next = tempNext;
		}
		return head;
	}
private:
	ListNode* swapAll(ListNode* head, ListNode *&end)
	{
		if (!head) return end = nullptr;
		if (!head->next) return end = head;
		end = head;
		ListNode* oldSecond;
		ListNode* newHead = swapAll(head->next, oldSecond);
		oldSecond->next = end;
		end->next = nullptr;
		return newHead;
	}
};

File: 25.reverse-nodes-in-k-group/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        p = head
        head = ListNode(0)
        head.next = p
        p = head
        while True:
            q = p
            for i in range(k):
                if p.next is None:
                    return head.next
                p = p.next
            new_head, new_tail = self.reverseJustKGroup(q.next, k)
            q.next = new_head
            p = new_tail

    def reverseJustKGroup(self, head: ListNode, k: int) -> (ListNode, ListNode):
        if k == 1:
            return head, head
        new_head, new_tail = self.reverseJustKGroup(head.next, k - 1)
        head.next = new_tail.next
        new_tail.next = head
        return new_head, new_tail.next
